### PR TITLE
Sign Windows binaries

### PR DESCRIPTION
**Pull request by blampe**
Port of https://github.com/pulumi/ci-mgmt/pull/1202.

The build step is re-used by the subsequent dist step, so overall clock time is unchanged around ~8m50s.

Fixes https://github.com/pulumi/pulumi-azure-native/issues/3642.

### CODE REVIEW

File: .github/workflows/build-test.yml [hunk 1]
```diff
@@ -366,6 +366,27 @@ jobs:
           make --touch codegen schema
           make provider_prebuild
 
+      - name: Build and sign windows provider
+        run: |
```
Comment (thomas11) [inline .github/workflows/build-test.yml:370]: Shouldn't we use `set -euxo pipefail` or similar here to detect issues and fail early?

File: .github/workflows/build-test.yml [hunk 2]
```diff
@@ -366,6 +366,27 @@ jobs:
...
+          make bin/windows-amd64/pulumi-resource-azure-native.exe;
+
+          az login --service-principal \
+            -u ${{ secrets.AZURE_SIGNING_CLIENT_ID }} \
```
Comment (thomas11) [inline .github/workflows/build-test.yml:374]: [nit] long `--` args are more readable in scripts

File: .github/workflows/build-test.yml [hunk 3]
```diff
@@ -366,6 +366,28 @@ jobs:
           make --touch codegen schema
           make provider_prebuild
 
+      - name: Build and sign windows provider
+        shell: bash # Runs with -eo pipefail
+        run: |
+          make bin/windows-amd64/pulumi-resource-azure-native.exe;
```
Comment (danielrbradley) [inline .github/workflows/build-test.yml:372]: So, this is building just the windows binary first, then relying on the `make dist` from skipping the rebuild as it should be up-to-date? 

Slightly more tempted to move this into the makefile given the implicit assumptions on the target associations here. Might be able to spike on this myself now you've done the legwork of the actual signing code.

Comment (danielrbradley) [reply .github/workflows/build-test.yml:372]: Quick hack of what the extra code in the platform-specific provider binary target might look like:

```makefile
ifeq ($(TARGET),windows-amd64)
ifeq (findstring ||,|${AZURE_SIGNING_CLIENT_ID}|${AZURE_SIGNING_CLIENT_SECRET}|${AZURE_SIGNING_TENANT_ID}|)
	echo "Skipping signing as Azure credentials are not set (AZURE_SIGNING_CLIENT_ID, AZURE_SIGNING_CLIENT_SECRET, AZURE_SIGNING_TENANT_ID)"
else
	az login --service-principal \
            --username ${{ secrets.AZURE_SIGNING_CLIENT_ID }} \
            --password ${{ secrets.AZURE_SIGNING_CLIENT_SECRET }} \
            --tenant ${{ secrets.AZURE_SIGNING_TENANT_ID }} \
            --output none && \
		wget https://github.com/ebourg/jsign/releases/download/6.0/jsign-6.0.jar && \
		java -jar jsign-6.0.jar \
             --storetype AZUREKEYVAULT \
             --keystore "PulumiCodeSigning" \
             --url ${{ secrets.AZURE_SIGNING_KEY_VAULT_URI }} \
             --storepass "$(az account get-access-token --resource "https://vault.azure.net" | jq -r .accessToken)" \
             bin/windows-amd64/pulumi-resource-azure-native.exe && \
		az logout
endif
endif
```

**Review by danielrbradley — APPROVED** [review]
Tested the makefile alternative for a little bit but can't get it working yet, so happy to ship this as-is.

File: .github/workflows/build-test.yml (hunk 3 above)
Comment (danielrbradley) [reply .github/workflows/build-test.yml:372]: Ok, new bash version seems like it's half way there ... can't test fully due to no cred though:

```makefile
bin/%/$(PROVIDER) bin/%/$(PROVIDER).exe: bin/pulumictl .make/provider_mod_download .make/prebuild provider/cmd/$(PROVIDER)/*.go .make/provider_prebuild $(PROVIDER_PKG)
	@# check the TARGET is set
	test $(TARGET)
	cd provider && \
		export GOOS=$$(echo "$(TARGET)" | cut -d "-" -f 1) && \
		export GOARCH=$$(echo "$(TARGET)" | cut -d "-" -f 2) && \
		CGO_ENABLED=0 go build -o ${WORKING_DIR}/$@ $(VERSION_FLAGS) $(PROJECT)/v2/provider/cmd/$(PROVIDER)
	if [[ "${TARGET}" = "windows-amd64" && "|${AZURE_SIGNING_CLIENT_ID}|${AZURE_SIGNING_CLIENT_SECRET}|${AZURE_SIGNING_TENANT_ID}|" == *"||"* ]]; then \
			echo "Skipping signing as Azure credentials are not set (AZURE_SIGNING_CLIENT_ID, AZURE_SIGNING_CLIENT_SECRET, AZURE_SIGNING_TENANT_ID)"; \
	else \
		az login --service-principal \
				--username ${AZURE_SIGNING_CLIENT_ID} \
				--password ${AZURE_SIGNING_CLIENT_SECRET} \
				--tenant ${AZURE_SIGNING_TENANT_ID} \
				--output none; \
		wget https://github.com/ebourg/jsign/releases/download/6.0/jsign-6.0.jar --output-document=jsign-6.0.jar; \
		java -jar jsign-6.0.jar \
			--storetype AZUREKEYVAULT \
			--keystore "PulumiCodeSigning" \
			--url ${AZURE_SIGNING_KEY_VAULT_URI} \
			--storepass "$$(az account get-access-token --resource "https://vault.azure.net" | jq -r .accessToken)" \
			bin/windows-amd64/pulumi-resource-azure-native.exe; \
		az logout; \
	fi
```

Comment (danielrbradley) [reply .github/workflows/build-test.yml:372]: Pushed as alternative PR here: https://github.com/pulumi/pulumi-azure-native/pull/3774

**Pull request merged**